From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Evan <john.doe@somewhere.on.planet>
Date: Wed, 4 Mar 2026 14:31:16 +0800
Subject: Patching kernel rk35xx files drivers/gpio/Kconfig
 drivers/gpio/Makefile drivers/gpio/gpio-nca9555.c drivers/mfd/Kconfig
 drivers/mfd/Makefile drivers/mfd/nca9555.c
 include/dt-bindings/gpio/nca9555-gpio.h include/linux/mfd/nca9555.h

Signed-off-by: John Doe <john.doe@somewhere.on.planet>
---
 drivers/gpio/Kconfig                    |   7 +
 drivers/gpio/Makefile                   |   1 +
 drivers/gpio/gpio-nca9555.c             | 192 ++++++++++
 drivers/mfd/Kconfig                     |   9 +
 drivers/mfd/Makefile                    |   1 +
 drivers/mfd/nca9555.c                   | 144 +++++++
 include/dt-bindings/gpio/nca9555-gpio.h |  29 ++
 include/linux/mfd/nca9555.h             |  42 ++
 8 files changed, 425 insertions(+)

diff --git a/drivers/gpio/Kconfig b/drivers/gpio/Kconfig
index 3bb79f930ae2..de59ce4a4b0e 100644
--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -123,10 +123,17 @@ config GPIO_74XX_MMIO
 	    2 bits:	742G125 (Input), 7474 (Output)
 	    4 bits:	74125 (Input), 74175 (Output)
 	    6 bits:	74365 (Input), 74174 (Output)
 	    8 bits:	74244 (Input), 74273 (Output)
 	    16 bits:	741624 (Input), 7416374 (Output)
+		
+config GPIO_NCA9555
+       tristate "GPIO Support for NCA9555"
+       depends on MFD_NCA9555
+       help
+         This option enables support for on-chip GPIO found
+         on Analog Devices NCA9555
 
 config GPIO_ALTERA
 	tristate "Altera GPIO"
 	depends on OF_GPIO
 	select GPIOLIB_IRQCHIP
diff --git a/drivers/gpio/Makefile b/drivers/gpio/Makefile
index 76445d0fb22a..156cd1bfa336 100644
--- a/drivers/gpio/Makefile
+++ b/drivers/gpio/Makefile
@@ -187,5 +187,6 @@ obj-$(CONFIG_GPIO_XLP)			+= gpio-xlp.o
 obj-$(CONFIG_GPIO_XRA1403)		+= gpio-xra1403.o
 obj-$(CONFIG_GPIO_XTENSA)		+= gpio-xtensa.o
 obj-$(CONFIG_GPIO_ZEVIO)		+= gpio-zevio.o
 obj-$(CONFIG_GPIO_ZYNQ)			+= gpio-zynq.o
 obj-$(CONFIG_GPIO_ZYNQMP_MODEPIN)	+= gpio-zynqmp-modepin.o
+obj-$(CONFIG_GPIO_NCA9555)      += gpio-nca9555.o
diff --git a/drivers/gpio/gpio-nca9555.c b/drivers/gpio/gpio-nca9555.c
new file mode 100755
index 000000000000..37d043677361
--- /dev/null
+++ b/drivers/gpio/gpio-nca9555.c
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * GPIO driver for NCA9555 MFD
+ *
+ * Copyright 2023 RPDZKJ
+ */
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/platform_device.h>
+#include <linux/mfd/nca9555.h>
+#include <linux/gpio/driver.h>
+
+#define NCA9555_GPIO_MAX 16
+
+struct nca9555_gpio_dev {
+	struct device *parent;
+	struct gpio_chip gpio_chip;
+	struct mutex lock;
+	u8 dat_out[2];
+	u8 dir[2];
+};
+
+static int nca9555_gpio_reg_read(struct nca9555_gpio_dev *nca9555_gpio, u8 reg, u8 *val)
+{
+	struct nca9555_dev *nca9555;
+	nca9555 = dev_get_drvdata(nca9555_gpio->parent);
+
+	return nca9555->read_reg(nca9555, reg, val);
+}
+
+static int nca9555_gpio_reg_write(struct nca9555_gpio_dev *nca9555_gpio, u8 reg, u8 val)
+{
+	struct nca9555_dev *nca9555;
+	nca9555 = dev_get_drvdata(nca9555_gpio->parent);
+
+	return nca9555->write_reg(nca9555, reg, val);
+}
+
+static int nca9555_gpio_get_value(struct gpio_chip *chip, unsigned int off)
+{
+	struct nca9555_gpio_dev *nca9555_gpio;
+	unsigned int bank, bit, ret;
+	u8 val;
+
+	nca9555_gpio = gpiochip_get_data(chip);
+	bank = NCA9555_BANK(off);
+	bit = NCA9555_BIT(off);
+
+	mutex_lock(&nca9555_gpio->lock);
+	/* There are dedicated registers for GPIO IN/OUT. */
+	if (nca9555_gpio->dir[bank] & bit)
+        ret=nca9555_gpio_reg_read(nca9555_gpio, GPIO_PORT0_INPUT_REG + bank, &val);
+	else	
+		val = nca9555_gpio->dat_out[bank];
+	mutex_unlock(&nca9555_gpio->lock);
+
+	return !!(val & bit);
+}
+
+static void nca9555_gpio_set_value(struct gpio_chip *chip, unsigned int off, int val)
+{
+	struct nca9555_gpio_dev *nca9555_gpio;
+	unsigned int bank, bit;
+
+	nca9555_gpio = gpiochip_get_data(chip);
+	bank = NCA9555_BANK(off);
+	bit = NCA9555_BIT(off);
+
+	mutex_lock(&nca9555_gpio->lock);
+	if (val)
+		nca9555_gpio->dat_out[bank] |= bit;
+	else
+		nca9555_gpio->dat_out[bank] &= ~bit;
+
+	nca9555_gpio_reg_write(nca9555_gpio, GPIO_PORT0_OUTPUT_REG + bank,
+			       nca9555_gpio->dat_out[bank]);
+	mutex_unlock(&nca9555_gpio->lock);
+}
+
+static int nca9555_gpio_direction_input(struct gpio_chip *chip, unsigned int off)
+{
+	struct nca9555_gpio_dev *nca9555_gpio;
+	unsigned int bank, bit;
+	int ret;
+
+	nca9555_gpio = gpiochip_get_data(chip);
+	bank = NCA9555_BANK(off);
+	bit = NCA9555_BIT(off);
+
+	mutex_lock(&nca9555_gpio->lock);
+	nca9555_gpio->dir[bank] |= bit;
+	ret = nca9555_gpio_reg_write(nca9555_gpio, GPIO_PORT0_CONFIG_REG + bank,
+				     nca9555_gpio->dir[bank]);
+	mutex_unlock(&nca9555_gpio->lock);
+	return ret;
+}
+
+static int nca9555_gpio_direction_output(struct gpio_chip *chip, unsigned int off, int val)
+{
+	struct nca9555_gpio_dev *nca9555_gpio;
+	unsigned int bank, bit;
+	int ret;
+
+	nca9555_gpio = gpiochip_get_data(chip);
+
+	bank = NCA9555_BANK(off);
+	bit = NCA9555_BIT(off);
+
+	mutex_lock(&nca9555_gpio->lock);
+	nca9555_gpio->dir[bank] &= ~bit;
+
+	if (val)
+		nca9555_gpio->dat_out[bank] |= bit;
+	else
+		nca9555_gpio->dat_out[bank] &= ~bit;
+	ret = nca9555_gpio_reg_write(nca9555_gpio, GPIO_PORT0_OUTPUT_REG + bank,
+				     nca9555_gpio->dat_out[bank]);
+	ret |= nca9555_gpio_reg_write(nca9555_gpio, GPIO_PORT0_CONFIG_REG + bank,
+				      nca9555_gpio->dir[bank]);
+
+	mutex_unlock(&nca9555_gpio->lock);
+
+	return ret;
+}
+
+static int nca9555_gpio_probe(struct platform_device *pdev)
+{
+	struct nca9555_gpio_dev *nca9555_gpio;
+	struct device *dev = &pdev->dev;
+	struct device_node *np = dev->of_node;
+	struct gpio_chip *gc;
+	int i;
+
+	dev_info(dev, "gpio-nca9555 probe start \n");
+
+	nca9555_gpio = devm_kzalloc(&pdev->dev, sizeof(struct nca9555_gpio_dev), GFP_KERNEL);
+	if (nca9555_gpio == NULL)
+		return -ENOMEM;
+
+	nca9555_gpio->parent = pdev->dev.parent;
+
+	gc = &nca9555_gpio->gpio_chip;
+	gc->of_node = np;
+	gc->parent = dev;
+	gc->direction_input  = nca9555_gpio_direction_input;
+	gc->direction_output = nca9555_gpio_direction_output;
+	gc->get = nca9555_gpio_get_value;
+	gc->set = nca9555_gpio_set_value;
+	gc->can_sleep = true;
+
+	gc->base = -1;
+	gc->ngpio = NCA9555_GPIO_MAX;
+	gc->label = pdev->name;
+	gc->owner = THIS_MODULE;
+
+	mutex_init(&nca9555_gpio->lock);
+
+	for (i = 0; i < 2; i++) {
+		u8 *dat_out, *dir;
+		dat_out = nca9555_gpio->dat_out;
+		dir = nca9555_gpio->dir;
+		nca9555_gpio_reg_read(nca9555_gpio,
+				      GPIO_PORT0_OUTPUT_REG + i, dat_out + i);
+		nca9555_gpio_reg_read(nca9555_gpio,
+				      GPIO_PORT0_CONFIG_REG + i, dir + i);
+	}
+	
+	dev_info(dev, "gpio-nca9555 probe succes \n");
+	return devm_gpiochip_add_data(&pdev->dev, &nca9555_gpio->gpio_chip, nca9555_gpio);
+}
+
+static const struct of_device_id nca9555_of_match[] = {
+	{.compatible = "nca9555-gpio", },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, nca9555_of_match);
+
+static struct platform_driver nca9555_gpio_driver = {
+	.driver	= {
+		.name	= "nca9555-gpio",
+		.of_match_table = nca9555_of_match,
+	},
+	.probe		= nca9555_gpio_probe,
+};
+
+module_platform_driver(nca9555_gpio_driver);
+
+MODULE_AUTHOR("Haibo Chen <haibo.chen@nxp.com>");
+MODULE_DESCRIPTION("GPIO NCA9555 Driver");
+MODULE_LICENSE("GPL");
diff --git a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
index 4fe9467404bd..60e7167bb085 100644
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -167,10 +167,19 @@ config MFD_BD9571MWV
 	  watchdog block.
 
 	  This driver can also be built as a module. If so, the module
 	  will be called bd9571mwv.
 
+config MFD_NCA9555
+        bool "Analog Devices NCA9555 MFD PMIC Core Support"
+        depends on I2C=y
+        help
+          Say yes here to add support for Analog Devices NCA9555,
+          Multifunction Power Management IC. This includes
+          the I2C driver and the core APIs _only_,
+          Support NCA9555 GPIO expander.
+
 config MFD_AC100
 	tristate "X-Powers AC100"
 	select MFD_CORE
 	depends on SUNXI_RSB
 	help
diff --git a/drivers/mfd/Makefile b/drivers/mfd/Makefile
index a9589498a27a..1fc4020c45f8 100644
--- a/drivers/mfd/Makefile
+++ b/drivers/mfd/Makefile
@@ -289,5 +289,6 @@ obj-$(CONFIG_MFD_INTEL_M10_BMC)   += intel-m10-bmc.o
 obj-$(CONFIG_MFD_ATC260X)	+= atc260x-core.o
 obj-$(CONFIG_MFD_ATC260X_I2C)	+= atc260x-i2c.o
 
 obj-$(CONFIG_MFD_RSMU_I2C)	+= rsmu_i2c.o rsmu_core.o
 obj-$(CONFIG_MFD_RSMU_SPI)	+= rsmu_spi.o rsmu_core.o
+obj-$(CONFIG_MFD_NCA9555)   += nca9555.o
\ No newline at end of file
diff --git a/drivers/mfd/nca9555.c b/drivers/mfd/nca9555.c
new file mode 100755
index 000000000000..a3d22529f14a
--- /dev/null
+++ b/drivers/mfd/nca9555.c
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/* NCA9555 IO Expander, Key controller core driver.
+ *
+ * Copyright 2023 RPDZKJ
+ */
+
+#include <linux/module.h>
+#include <linux/moduleparam.h>
+#include <linux/init.h>
+#include <linux/slab.h>
+#include <linux/i2c.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/mfd/core.h>
+#include <linux/mfd/nca9555.h>
+
+static const struct mfd_cell nca9555_devs[] = {
+	{
+		.name = "nca9555-gpio",
+		.of_compatible = "nca9555-gpio",
+	}
+};
+
+static int nca9555_i2c_read_reg(struct nca9555_dev *nca9555, u8 reg, u8 *val)
+{
+	struct i2c_client *i2c = nca9555->i2c_client;
+	struct i2c_msg xfer[2];
+	int ret;
+
+	/* Write register */
+	xfer[0].addr = i2c->addr;
+	xfer[0].flags = 0;
+	xfer[0].len = 1;
+	xfer[0].buf = &reg;
+
+	/* Read data */
+	xfer[1].addr = i2c->addr;
+	xfer[1].flags = I2C_M_RD;
+	xfer[1].len = 1;
+	xfer[1].buf = val;
+
+	ret = i2c_transfer(i2c->adapter, xfer, 2);
+	if (ret == 2) {
+		return 0;
+	} else {
+		dev_err(&i2c->dev,
+			"Failed to read reg 0x%02x, ret is %d\n",
+			reg, ret);
+		if ( ret >=0)
+			return -EIO;
+		else
+			return ret;
+	}
+}
+
+static int nca9555_i2c_write_reg(struct nca9555_dev *nca9555, u8 reg, u8 val)
+{
+	struct i2c_client *i2c = nca9555->i2c_client;
+	u8 msg[2];
+	int ret;
+
+	msg[0] = reg;
+	msg[1] = val;
+	ret = i2c_master_send(i2c, msg, 2);
+	if (ret == 2) {
+		return 0;
+	} else {
+		dev_err(&i2c->dev,
+			"Failed to write reg 0x%02x, ret is %d\n",
+			reg, ret);
+		if ( ret >=0)
+			return -EIO;
+		else
+			return ret;
+	}
+}
+
+static int nca9555_i2c_probe(struct i2c_client *i2c,
+				const struct i2c_device_id *id)
+{
+	struct nca9555_dev *nca9555;
+	u8 reg;
+	int ret;
+
+	dev_info(&i2c->dev, "nca9555 probe start \n");
+
+	nca9555 = devm_kzalloc(&i2c->dev, sizeof(struct nca9555_dev),
+				GFP_KERNEL);
+	if (nca9555 == NULL)
+		return -ENOMEM;
+
+	i2c_set_clientdata(i2c, nca9555);
+	nca9555->dev = &i2c->dev;
+	nca9555->i2c_client = i2c;
+	nca9555->read_reg = nca9555_i2c_read_reg;
+	nca9555->write_reg = nca9555_i2c_write_reg;
+
+	ret = nca9555_i2c_read_reg(nca9555, NCA9555_ID, &reg);
+	if (ret)
+		return ret;
+
+	dev_info(&i2c->dev, "nca9555 probe success \n");
+	return devm_mfd_add_devices(nca9555->dev, PLATFORM_DEVID_AUTO,
+				    nca9555_devs, ARRAY_SIZE(nca9555_devs),
+				    NULL, 0, NULL);
+}
+
+static const struct i2c_device_id nca9555_i2c_id[] = {
+	{ "nca9555", 0 },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(i2c, nca9555_i2c_id);
+
+static const struct of_device_id nca9555_of_match[] = {
+	{.compatible = "nca9555", },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, nca9555_of_match);
+
+static struct i2c_driver nca9555_i2c_driver = {
+	.driver = {
+		   .name = "nca9555",
+		   .of_match_table = of_match_ptr(nca9555_of_match),
+	},
+	.probe = nca9555_i2c_probe,
+	.id_table = nca9555_i2c_id,
+};
+
+static int __init nca9555_i2c_init(void)
+{
+	return i2c_add_driver(&nca9555_i2c_driver);
+}
+/* init early so consumer devices can complete system boot */
+subsys_initcall(nca9555_i2c_init);
+
+static void __exit nca9555_i2c_exit(void)
+{
+	i2c_del_driver(&nca9555_i2c_driver);
+}
+module_exit(nca9555_i2c_exit);
+
+MODULE_DESCRIPTION("NCA9555 core driver");
+MODULE_AUTHOR("Haibo Chen <haibo.chen@nxp.com>");
+MODULE_LICENSE("GPL");
diff --git a/include/dt-bindings/gpio/nca9555-gpio.h b/include/dt-bindings/gpio/nca9555-gpio.h
new file mode 100755
index 000000000000..62fead2db33c
--- /dev/null
+++ b/include/dt-bindings/gpio/nca9555-gpio.h
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * GPIO definitions for NCA9555
+ *
+ * Copyright (C) 2023 RPDZKJ
+ */
+
+#ifndef _DT_BINDINGS_NCA9555_GPIO_H
+#define _DT_BINDINGS_NCA9555_GPIO_H
+
+#define IO_00  0
+#define IO_01  1
+#define IO_02  2
+#define IO_03  3
+#define IO_04  4
+#define IO_05  5
+#define IO_06  6
+#define IO_07  7
+
+#define IO_10  8
+#define IO_11  9
+#define IO_12  10
+#define IO_13  11
+#define IO_14  12
+#define IO_15  13
+#define IO_16  14
+#define IO_17  15
+
+#endif /* _DT_BINDINGS_NCA9555_GPIO_H */
diff --git a/include/linux/mfd/nca9555.h b/include/linux/mfd/nca9555.h
new file mode 100755
index 000000000000..f41bab4dcb5c
--- /dev/null
+++ b/include/linux/mfd/nca9555.h
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Analog Devices NCA9555 I/O Expander, keypad controller,
+ * PWM contorller.
+ *
+ * Copyright 2022 NXP
+ */
+
+#ifndef __NCA9555_H_
+#define __NCA9555_H_
+
+
+#define GPIO_PORT0 0
+#define GPIO_PORT1 1
+
+#define GPIO_PORT0_INPUT_REG     0x00
+#define GPIO_PORT1_INPUT_REG     0x01
+
+#define GPIO_PORT0_OUTPUT_REG    0x02
+#define GPIO_PORT1_OUTPUT_REG    0x03
+
+#define GPIO_PORT0_POLAR_REG     0x04
+#define GPIO_PORT1_POLAR_REG     0x05
+
+#define GPIO_PORT0_CONFIG_REG    0x06
+#define GPIO_PORT1_CONFIG_REG    0x07
+
+#define NCA9555_ID			0x00
+
+#define NCA9555_BANK(offs)		((offs) >= 8)
+#define NCA9555_BIT(offs)		(offs >= 8 ? \
+					1u << (offs - 8) : 1u << (offs))
+
+struct nca9555_dev {
+	struct device *dev;
+	struct i2c_client *i2c_client;
+
+	int (*read_reg)(struct nca9555_dev *nca9555, u8 reg, u8 *val);
+	int (*write_reg)(struct nca9555_dev *nca9555, u8 reg, u8 val);
+};
+
+#endif
-- 
Created with Armbian build tools https://github.com/armbian/build

